### PR TITLE
Document precompiled Groovy script plugins

### PR DESCRIPTION
**Pull request by rieske**
To view the result visually: `./gradlew serveDocs -x userguideSinglePage`

Moved Kotlin precompiled scripts documentation from [Kotlin DSL primer](http://localhost:8000/userguide/kotlin_dsl.html) to [Developing Custom Gradle plugins](http://localhost:8000/userguide/custom_plugins.html#sec:precompiled_plugins) and updated the examples to include Groovy counterparts.

Added a release note about this new feature with a pointer to the [userguide section](http://localhost:8000/userguide/custom_plugins.html#sec:precompiled_plugins)

Updated the [precompiled plugins sample](http://localhost:8000/samples/sample_precompiled_script_plugin.html) to include Groovy version of precompiled plugins.

**Important**
This change also fixes a critical bug with applying other plugins within a precompiled Groovy plugin.
This was revealed within a documentation snippet test - however many plugins are added in a precompiled Groovy plugin, only one of them would be actually applied.

### CODE REVIEW

File: subprojects/docs/src/docs/userguide/api/kotlin_dsl.adoc [hunk 1]
```diff
@@ -740,55 +740,6 @@ You can silence the warning by setting the `experimentalWarning` property of the
 include::sample[dir="snippets/kotlinDsl/kotlinDslPlugin",files="buildSrc/build.gradle.kts[tags=disable-experimental-warning]"]
 ====
 
-[[kotdsl:precompiled_plugins]]
```
Comment (eskatos) [inline subprojects/docs/src/docs/userguide/api/kotlin_dsl.adoc:743]: ❓ I think we should add some http redirection on `docs.gradle.org` to support existing direct links to this after the move.

Hey @eriwen, do you still manage those? if not, who could help?

File: subprojects/docs/src/docs/release/notes.md [hunk 2]
```diff
@@ -43,6 +43,19 @@ It is however stable and expected to become the default format in Gradle 7.0.
 
 Take a look at [the documentation](userguide/dependency_locking.html#single_lock_file_per_project) for more information and how to enable the feature.
 
+## Precompiled Groovy script plugins
+
+Gradle now allows precompiled script plugins to be written in Groovy in addition to already present Kotlin DSL support.
```
Comment (big-guy) [inline subprojects/docs/src/docs/release/notes.md:48]:
```suggestion
Gradle now allows precompiled script plugins to be written using the Groovy DSL in addition to the Kotlin DSL.
```

File: subprojects/docs/src/docs/release/notes.md [hunk 3]
```diff
@@ -43,6 +43,19 @@ It is however stable and expected to become the default format in Gradle 7.0.
...
+
+Precompiled script plugins are binary plugins that are written in one of Gradle's DSLs.
```
Comment (big-guy) [inline subprojects/docs/src/docs/release/notes.md:50]:
```suggestion
Precompiled script plugins are binary plugins that are written in one of Gradle's DSL languages and look like regular build scripts.
```

File: subprojects/docs/src/docs/release/notes.md [hunk 4]
```diff
@@ -43,6 +43,19 @@ It is however stable and expected to become the default format in Gradle 7.0.
...
+They can be resolved and applied using the `plugins {}` block. They can also be published and shared just like regular binary plugins.
+
+For example, a Gradle script in `buildSrc/src/main/groovy/my-plugin.gradle` can be used as a plugin in the main project as `plugins { id 'my-plugin'}`.
```
Comment (big-guy) [inline subprojects/docs/src/docs/release/notes.md:53]:
```suggestion
For example, a Gradle script in `buildSrc/src/main/groovy/my-plugin.gradle` can be used as a plugin in the main project as `plugins { id 'my-plugin' }`.
```

File: subprojects/docs/src/docs/release/notes.md [hunk 5]
```diff
@@ -43,6 +43,19 @@ It is however stable and expected to become the default format in Gradle 7.0.
...
+
+Precompiled script plugins are covered in more depth in the [user guide](userguide/custom_plugins.html#sec:precompiled_plugins).
```
Comment (big-guy) [inline subprojects/docs/src/docs/release/notes.md:55]: We try to use 'user manual' over 'user guide' (even though the url is still userguide).

```suggestion
Precompiled script plugins are covered in more depth in the [user manual](userguide/custom_plugins.html#sec:precompiled_plugins).
```

File: subprojects/docs/src/docs/userguide/api/kotlin_dsl.adoc (hunk 1 above)
Comment (big-guy) [reply subprojects/docs/src/docs/userguide/api/kotlin_dsl.adoc:743]: @eskatos I don't think we can fix anchors with a redirect.  We can only redirect to an entirely different page if this page moved.

All of the old docs will still work, so this should only be a problem if we have non-versioned links to this.  If you think this is a serious problem, we can leave the anchor here and just refer to the new chapter/section as a link. WDYT?

File: subprojects/docs/src/docs/userguide/extending-gradle/custom_plugins.adoc [hunk 6]
```diff
@@ -231,8 +231,89 @@ include::sample[dir="snippets/customPlugin/groovy/consumer",files="settings.grad
 include::sample[dir="snippets/customPlugin/kotlin/consumer",files="settings.gradle.kts[tags=use-legacy-plugin]"]
 ====
 
+[[sec:precompiled_plugins]]
+== Precompiled script plugins
+In addition to plugins written as regular projects, Gradle allows you to provide the plugin logic
+in either Groovy or Kotlin DSL as precompiled script plugins. You write these as `\*.gradle` files in `src/main/groovy`
+directory or `*.gradle.kts` files in `src/main/kotlin` directory.
```
Comment (big-guy) [inline subprojects/docs/src/docs/userguide/extending-gradle/custom_plugins.adoc:238]:
```suggestion
In addition to plugins written as standalone projects, Gradle also allows you to provide build logic written in either Groovy or Kotlin DSLs as precompiled script plugins. You write these as `\*.gradle` files in `src/main/groovy`
directory or `*.gradle.kts` files in `src/main/kotlin` directory.
```

File: subprojects/docs/src/docs/userguide/extending-gradle/custom_plugins.adoc [hunk 7]
```diff
@@ -231,8 +231,89 @@ include::sample[dir="snippets/customPlugin/groovy/consumer",files="settings.grad
...
+
+Precompiled script plugins are compiled as part of a regular source set and placed on the build classpath or packaged in
+a binary plugin, depending on what type of project they are in. For all intents and purposes, they are binary plugins,
+particularly as they can be applied by plugin ID, just like a regular binary plugin. In fact, the plugin metadata for
+them is generated via implicitly applied <<java_gradle_plugin#java_gradle_plugin,Gradle Plugin Development Plugin>>.
```
Comment (big-guy) [inline subprojects/docs/src/docs/userguide/extending-gradle/custom_plugins.adoc:243]:
```suggestion
Precompiled script plugins are compiled into class files and packaged into a jar. For all intents and purposes, they are binary plugins and can be applied by plugin ID, tested and published as binary plugins. In fact, the plugin metadata for
them is generated using the <<java_gradle_plugin#java_gradle_plugin,Gradle Plugin Development Plugin>>.
```

File: subprojects/docs/src/docs/userguide/extending-gradle/custom_plugins.adoc [hunk 8]
```diff
@@ -231,8 +231,89 @@ include::sample[dir="snippets/customPlugin/groovy/consumer",files="settings.grad
...
+
+[NOTE]
+====
+Kotlin DSL precompiled script plugins built with Gradle 6.0 cannot be used with earlier versions of Gradle.
+This limitation will be lifted in a future version of Gradle.
+
+Groovy DSL precompiled script plugins are available starting with Gradle 6.4 and they can not be used with earlier versions.
```
Comment (big-guy) [inline subprojects/docs/src/docs/userguide/extending-gradle/custom_plugins.adoc:250]: I think it's OK to say this is the limitation for Groovy DSL until we're sure what the backwards guarantees will be in the future.

**Pull request merged**